### PR TITLE
Added variable latency delay, normal and uniform based

### PR DESCRIPTION
**Pull request by rikonor**
Hey,
I saw the todo for random delays and thought I'd give it a go.

License: MIT
Signed-off-by: Or rikonor@gmail.com

### CODE REVIEW

File: thirdparty/delay/delay.go [hunk 1]
```diff
@@ -44,3 +43,53 @@ func (d *delay) Get() time.Duration {
...
+type variableUniform struct {
+	delay
+	d   time.Duration // max delta
+	rng *rand.Rand
+}
+
+func (d *variableUniform) Wait() {
+	d.l.RLock()
+	defer d.l.RUnlock()
+	randomDelay := time.Duration(d.rng.Float64() * float64(d.d))
```
Comment (Kubuxu) [inline thirdparty/delay/delay.go]: I think unlock should be here so multiple goroutines can Wait at the same time.

File: thirdparty/delay/delay.go [hunk 2]
```diff
@@ -44,3 +43,53 @@ func (d *delay) Get() time.Duration {
...
 	return d.t
 }
+
+// VariableNormal is a delay following a normal distribution
+// Notice that to implement the D interface Set can only change the mean delay
+// the standard deviation is set only at initialization
+func VariableNormal(t, std time.Duration) D {
+	v := &variableNormal{
+		std: std,
+		rng: rand.New(rand.NewSource(int64(time.Now().Nanosecond()))),
```
Comment (Kubuxu) [inline thirdparty/delay/delay.go]: Using one random source per one delay instance runs risk of it being used just one time.
Pseudo Random Generators used just one time will not generate uniformly distributed random numbers.

I would recommend using one package wide random generator.

**Comment by rikonor** [comment]
@Kubuxu, thanks for the feedback.
I added a shared rng instance, and left the possibility of providing your own.
Regarding the locks, two goroutines can wait at the same time because they'll both be able to acquire a read lock.

**Comment by Kubuxu** [comment]
Right, I missed that.
Would you mind amending triangular brackets around you events just to make all tests happy.

Apart from that, LGTM.

**Comment by rikonor** [comment]
@Kubuxu - yep, fixed that. Thanks!

**Comment by Kubuxu** [comment]
LGTM

File: thirdparty/delay/delay.go [hunk 3]
```diff
@@ -1,10 +1,13 @@
 package delay
 
 import (
+	"math/rand"
 	"sync"
 	"time"
 )
 
+var sharedRNG = rand.New(rand.NewSource(int64(time.Now().Nanosecond())))
```
Comment (whyrusleeping) [inline thirdparty/delay/delay.go]: instead of `.Nanosecond()` we could use `.UnixNano()`, saves a cast

**Comment by whyrusleeping** [comment]
one comment, then LGTM

**Comment by rikonor** [comment]
@whyrusleeping, using `UnixNano` now instead of `Nanosecond` and a cast.

**Comment by whyrusleeping** [comment]
cool, LGTM. Thanks!

**Comment by lgierth** [comment]
This doesn't seem like it's used anywhere?

**Comment by rikonor** [comment]
@lgierth I don't believe that it is. I was just looking for something to do and ended up implementing that. Hope that it can help at some point in the future.

**Comment by Kubuxu** [comment]
We currently do local network dialups in bursts. This is somewhere we can use it.

**Comment by jbenet** [comment]
Maybe this type of thing could be useful as a libp2p transport?

**Comment by Kubuxu** [comment]
We might move it to utility package to use it in libp2p to not send dials in bursts, it isn't huge issues right now.

**Comment by lgierth** [comment]
@jbenet aren't you having a day off? Go out and play :)

**Pull request merged**